### PR TITLE
Dev zhi fix lsl hang2.0

### PR DESCRIPTION
**Pull request by zhili1208**

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 1]
```diff
@@ -576,24 +571,25 @@ private void OnEnvDTEProjectRemoved(EnvDTE.Project envDTEProject)
...
                     && !EnvDTEProjectUtility.IsParentProjectExplicitlyUnsupported(envDTEProject)
                     && _solutionOpenedRaised)
-                {
-                    await EnsureNuGetAndVsProjectAdapterCacheAsync();
-                    var vsProjectAdapter = await _vsProjectAdapterProvider.CreateAdapterForFullyLoadedProjectAsync(envDTEProject);
-                    AddVsProjectAdapterToCache(vsProjectAdapter);
-                    NuGetProject nuGetProject;
-                    _projectSystemCache.TryGetNuGetProject(envDTEProject.Name, out nuGetProject);
+                    {
+                        await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: Seemingly redundant. Consider removing.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 2]
```diff
@@ -757,63 +753,70 @@ private void RemoveVsProjectAdapterFromCache(string name)
...
-                        await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-
-                        await InitializeAsync();
-
-                        var dte = _serviceProvider.GetDTE();
-                        if (dte.Solution.IsOpen)
+                        if (context == VSSolutionManagerInitContext.SolutionOpen)
                         {
-                            OnSolutionExistsAndFullyLoaded();
+                            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: move it into the called method

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 3]
```diff
@@ -757,63 +753,70 @@ private void RemoveVsProjectAdapterFromCache(string name)
...
         }
 
-        private void EnsureInitialize()
+        private async Task EnsureInitializeAsync(VSSolutionManagerInitContext context = VSSolutionManagerInitContext.Default)
         {
-            try
-            {
-                // If already initialized, need not be on the UI thread
-                if (!_initialized)
+            await _nuGetLockService.ExecuteNuGetOperationAsync(
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: Replace with local SemaphoreSlim.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 4]
```diff
@@ -757,63 +753,70 @@ private void RemoveVsProjectAdapterFromCache(string name)
...
-                {
-                    // Check if the cache is initialized.
-                    // It is possible that the cache is not initialized, since,
-                    // the solution was not saved and/or there were no projects in the solution
-                    if (!_cacheInitialized && _solutionOpenedRaised)
-                    {
-                        NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
+                        else if (context == VSSolutionManagerInitContext.SolutionChange)
                         {
                             await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: same

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 5]
```diff
@@ -757,63 +753,70 @@ private void RemoveVsProjectAdapterFromCache(string name)
...
-                }
-                else
-                {
-                    // Check if the cache is initialized.
-                    // It is possible that the cache is not initialized, since,
-                    // the solution was not saved and/or there were no projects in the solution
-                    if (!_cacheInitialized && _solutionOpenedRaised)
-                    {
-                        NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
+                        else if (context == VSSolutionManagerInitContext.SolutionChange)
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: SolutionChange scenario does not need lock?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 6]
```diff
@@ -33,6 +34,7 @@ namespace NuGet.PackageManagement.VisualStudio
     public sealed class VSSolutionManager : IVsSolutionManager, IVsSelectionEvents
     {
         private static readonly INuGetProjectContext EmptyNuGetProjectContext = new EmptyNuGetProjectContext();
+        private readonly VSSolutionInitLock _initLock = new VSSolutionInitLock();
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: use `NuGetLockService` instead

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 7]
```diff
@@ -1025,5 +1023,49 @@ public IEnumerable<IVsProjectAdapter> GetAllVsProjectAdapters()
         }
 
         #endregion
+
+        private sealed class VSSolutionInitLock
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: remove it

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 8]
```diff
@@ -1025,5 +1023,49 @@ public IEnumerable<IVsProjectAdapter> GetAllVsProjectAdapters()
...
+                }
+                else
+                {
+                    await action();
+                }
+            }
+        }
+    }
+
+    public enum VSSolutionManagerInitContext
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: => private InitMode

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 9]
```diff
@@ -1025,5 +1023,49 @@ public IEnumerable<IVsProjectAdapter> GetAllVsProjectAdapters()
...
+    {
+        Default,
+        SolutionOpen,
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: => OnSolutionOpen

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/WorkspaceProjectBuildProperties.cs [hunk 10]
```diff
@@ -45,6 +45,8 @@ public string GetPropertyValue(string propertyName)
         {
             Assumes.NotNullOrEmpty(propertyName);
 
+            await TaskScheduler.Default;
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/WorkspaceProjectBuildProperties.cs:48]: add a comment

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProjectProvider.cs [hunk 11]
```diff
@@ -59,35 +59,29 @@ public sealed class LegacyPackageReferenceProjectProvider : INuGetProjectProvide
...
+        public async Task<NuGetProject> TryCreateNuGetProjectAsync(
             IVsProjectAdapter vsProjectAdapter,
             ProjectProviderContext _,
-            bool forceProjectType,
-            out NuGetProject result)
+            bool forceProjectType)
         {
             Assumes.Present(vsProjectAdapter);
 
             _threadingService.ThrowIfNotOnUIThread();
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProjectProvider.cs]: replace with Switch...

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/MSBuildNuGetProjectProvider.cs [hunk 12]
```diff
@@ -37,22 +38,19 @@ internal class MSBuildNuGetProjectProvider : INuGetProjectProvider
...
             IVsProjectAdapter vsProjectAdapter,
             ProjectProviderContext context,
-            bool forceProjectType,
-            out NuGetProject result)
+            bool forceProjectType)
         {
             Assumes.Present(vsProjectAdapter);
             Assumes.Present(context);
 
             _threadingService.ThrowIfNotOnUIThread();
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/MSBuildNuGetProjectProvider.cs]: change to switch

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/MSBuildNuGetProjectProvider.cs [hunk 13]
```diff
@@ -61,14 +59,12 @@ internal class MSBuildNuGetProjectProvider : INuGetProjectProvider
             // Project folder path is the packages config folder path
             var packagesConfigFolderPath = vsProjectAdapter.ProjectDirectory;
 
-            result = new VsMSBuildNuGetProject(
+            return await System.Threading.Tasks.Task.FromResult(new VsMSBuildNuGetProject(
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/MSBuildNuGetProjectProvider.cs]: unroll this line change

File: src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs [hunk 14]
```diff
@@ -185,18 +185,20 @@ public NuGetPackage()
...
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
             {
-                windowFrame.SetProperty((int)__VSFPROPID.VSFPROPID_OwnerCaption, string.Format(
-                    CultureInfo.CurrentCulture,
-                    Resx.Label_NuGetWindowCaption,
-                    project.ProjectName));
-            }
+                ThreadHelper.ThrowIfNotOnUIThread();
+
+                var project = await SolutionManager.GetVsProjectAdapterAsync(await SolutionManager.GetNuGetProjectSafeNameAsync(e.NuGetProject));
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs]: null-check

File: src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs [hunk 15]
```diff
@@ -475,51 +477,52 @@ private void PowerConsoleService_ExecuteEnd(object sender, EventArgs e)
...
-                        ? Resources.NoProjectSelected
-                        : string.Format(CultureInfo.CurrentCulture, Resources.DTE_ProjectUnsupported, projectName);
-
-                    MessageHelper.ShowWarningMessage(errorMessage, Resources.ErrorDialogBoxTitle);
-                }
-            });
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(
+                   async () =>
+                   {
+                       ThreadHelper.ThrowIfNotOnUIThread();
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs]: keep switch..

File: src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs [hunk 16]
```diff
@@ -739,30 +742,29 @@ private void BeforeQueryStatusForAddPackageDialog(object sender, EventArgs args)
...
-                // Enable the 'Manage NuGet Packages For Solution' dialog menu
-                // - if the console is NOT busy executing a command, AND
-                // - if the solution exists and not debugging and not building AND
-                // - if there are NuGetProjects. This means there are loaded, supported projects.
-                command.Enabled =
-                    IsSolutionExistsAndNotDebuggingAndNotBuilding() &&
-                    !ConsoleStatus.Value.IsBusy &&
-                    SolutionManager.GetNuGetProjects().Any();
-            });
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs:754]: verify context menu is disabled when build is running.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs [hunk 17]
```diff
@@ -269,30 +269,33 @@ public override void CleanUp()
         // Creates the project lists. Also called after a project is added/removed/renamed.
         private void CreateProjectLists()
         {
-            // unhook event handler
-            if (Projects != null)
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: make it async?

**Comment by alpaix** [comment]
Review summary. This PR imposes big risk on product reliability.
Major concerns:
- 🛑 Using of unawaited tasks created with `RunAsync`. Every task instance should be wrapped with `try-catch` clause.
- 🛑 Delayed event handling may lead to invalid state.
- 🛑 Behavior change in handling UI commands. This may break automation scenarios.
- 🛑 Delayed handling of `BeforeQuery...` may lead to incorrect context menu state rendering.
- ⚠️ Initialization lock looks better now. Needs more testing
- ⚠️ Overall insufficient testing.

**Comment by nkolev92** [comment]
@zhili1208 closing as redudant.

**Pull request closed without merging**